src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs

```diff
@@ -3149,12 +3148,14 @@ private static bool AttributeTypeNamesMatch(Type attributeType, Type reflectedAt
                 MethodInfo[] methods = eventSourceType.GetMethods(BindingFlags.DeclaredOnly | BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance);
                 EventAttribute defaultEventAttribute;
                 int eventId = 1;        // The number given to an event that does not have a explicitly given ID.
-                EventMetadata[]? eventData = null;
+                Dictionary<int, EventMetadata>? eventData = null;
                 Dictionary<string, string>? eventsByName = null;
                 if (source != null || (flags & EventManifestOptions.Strict) != 0)
                 {
-                    eventData = new EventMetadata[methods.Length + 1];
-                    eventData[0].Name = "";         // Event 0 is the 'write messages string' event, and has an empty name.
+                    eventData = new();
+                    EventMetadata newEventMetadata = default(EventMetadata);
+                    newEventMetadata.Name = ""; // Event 0 is the 'write messages string' event, and has an empty name.
+                    eventData[0] = newEventMetadata;
```

Comment:
```suggestion
                    ref EventMetadata newEventMetadata = CollectionsMarshal.GetValueRefOrAddDefault(eventData, 0, out _);
                    newEventMetadata.Name = ""; // Event 0 is the 'write messages string' event, and has an empty name.
```
Nit: You can use `CollectionsMarshal.GetValueRefOrAddDefault` here as well given that it is used in the other place. Avoids copy of the large EventMetadata struct, and it is less code.